Add ``QkParam`` to the C API

doichanj opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Adding C-API functions for Parameter for parameterized circuits


### Details and comments

This implements 
- create parameter from symbol
- free parameter
- Binary operations : add/sub/mul/div/pow
- Unary operations : sin/cos/tan/asin/acos/atan/log/exp/conjugate/abs/sign/neg
- bind/substitute parameters
- evaluate as real

doichanj committed:
Add C-API for ParameterExpression

Cryoris commented on crates/cext/src/parameter_expression.rs:
```diff
@@ -0,0 +1,930 @@
+/// name must be a pointer to valid C string to name of symbol
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_parameter_new(name: *const c_char) -> *mut Param {
```
Could we rename all functions to `qk_param_<...>` to reflect they act on the `QkParam` type?

doichanj replied:
renamed in [be08769](https://github.com/Qiskit/qiskit/pull/14837/commits/be08769b11c1a0cbc8de77de55e54dba9ea627cb)

Cryoris commented on crates/cext/src/parameter_expression.rs:
```diff
@@ -0,0 +1,930 @@
+///
+/// @param name The name of symbol.
+///
+/// @return A pointer to the created ParameterExpression.
```
Many places talk about `ParameterExpression`, these should be updated to talk about `QkParam` instead

doichanj replied:
replaced in [be08769](https://github.com/Qiskit/qiskit/pull/14837/commits/be08769b11c1a0cbc8de77de55e54dba9ea627cb)

Cryoris commented on crates/cext/src/parameter_expression.rs:
```diff
@@ -0,0 +1,930 @@
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_parameter_new(name: *const c_char) -> *mut Param {
+    let name = unsafe { CStr::from_ptr(name).to_str().unwrap() };
```
Every `unsafe` statement needs a safety comment explaining why it is safe to be called, e.g. here
```suggestion
    // SAFETY: Per documentation, ``name`` is a valid pointer to a C string
    let name = unsafe { CStr::from_ptr(name).to_str().unwrap() };
```

Cryoris commented on crates/cext/src/parameter_expression.rs:
```diff
@@ -0,0 +1,930 @@
+///
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub extern "C" fn qk_parameter_from_value(value: f64) -> *mut Param {
```
Could we call this `from_double` since the C input value will be a `double`? "value" seems ambiguous

doichanj replied:
renamed in [be08769](https://github.com/Qiskit/qiskit/pull/14837/commits/be08769b11c1a0cbc8de77de55e54dba9ea627cb)

Cryoris commented on crates/cext/src/parameter_expression.rs:
```diff
@@ -0,0 +1,930 @@
+///
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub extern "C" fn qk_parameter_from_complex(value: Complex64) -> *mut Param {
```
I'm not sure we need this function yet -- currently parameters are only supported in the circuit in the Rust data model, so they can only ever be real. Once we have parameters in the `SparseObservable` we would need it, but I would just skip this for now

Cryoris reviewed (commented):
In the current model of this PR we're allocating a new `QkParam` object per function call, which matches our Rust model. But I'm thinking we might want to be more explicit about memory allocations in our C interface and allow users to re-use allocations. For example, we could define the interface as
```c
QkParam qk_param_new(const char*); // as before

// Compute out = sin(in), plus allow returning an exit code.
QkExitCode qk_param_sin(QkParam* out, const QkParam* in);
// Compute out = left + right;
QkExitCode qk_param_add(QkParam* out, const QkParam *left, QkParam *right);
```
This has the advantages that users could re-use allocations (e.g. pass the same pointer as `in` and `out`) and we could return proper exit codes. It also matches the explicit allocation scheme we e.g. have in the `qk_transpile` function. The disadvantage is that its a bit more explicit to spell out, but the C API should give maximum control in my opinion and nice interfaces would be part of the C++ API that is built on top.

doichanj committed:
Merge remote-tracking branch 'upstream/main' into parameter-c-api

doichanj committed:
add test cases for Parameter C-API

doichanj committed:
rename qk_paramter_ to qk_param_, fix array format of bind/subs functions

doichanj committed:
fix test for param

doichanj committed:
use existing QkParam to save results of operations

doichanj committed:
fix test_param.c

doichanj committed:
fix test_param.c

doichanj committed:
fix for lint, add releasenote

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+            ParameterExpression::from_f64(0.),
+        ))))
+    } else {
+        let name = unsafe { CStr::from_ptr(name).to_str().unwrap() };
```
This needs a `// SAFETY` comment

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+/// name must be a pointer to valid C string to name of symbol
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_param_new(name: *const c_char) -> *mut Param {
```
Since there's not a single standard way of creating a `Param`, could we call this `qk_param_new_symbol` to indicate this will create a new free parameter? (in contrast to a `Param` that is a numeric value)

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+use qiskit_circuit::parameter::symbol_expr::{Symbol, SymbolExpr, Value};
+
+/// @ingroup QkParam
+/// Construct a new Parameter with a name of symbol
```
```suggestion
/// Construct a new ``QkParam`` representing an unbound symbol.
```

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+///
+/// # Safety
+///
+/// name must be a pointer to valid C string to name of symbol
```
We should make sure that these are all proper sentences 🙂 You could e.g. check existing safety comments to ensure they are correct & consistent. E.g. from the `qk_quantum_register_new` function:
```suggestion
/// The `name` parameter must be a pointer to memory that contains a valid
/// nul terminator at the end of the string. It also must be valid for reads of
/// bytes up to and including the nul terminator.
```

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_param_new(name: *const c_char) -> *mut Param {
+    if name.is_null() {
```
I would suggest we just raise an error if the pointer is null: this function has only the single purpose of creating a free symbol 🙂 (same for the case below where you check whether the string is empty)

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+}
+
+/// @ingroup QkParam
+/// Free the QkParam.
```
These should all use 
```suggestion
/// Free the ``QkParam``.
```

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+/// # Safety
+///
+/// Behavior is undefined if ``expr`` is not either null or a valid pointer to a
+/// [ParameterExpression].
```
The `[MyType]` links are unfortunately not going to be resolved in the docs, we have to use the names of the structs that they will have in the C API, i.e. 
```suggestion
/// ``QkParam``.
```

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+    Box::into_raw(Box::new(param))
+}
+
+// @ingroup QkParam
```
Could we remove these comments? If #14686 is merged adding this in seems simple enough 🙂

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_param_copy(expr: *const Param) -> *mut Param {
+    let expr = unsafe { const_ptr_as_ref(expr) };
```
These always need a `// SAFETY` comment 🙂

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+///
+///     QkParam *a = qk_param_new("a");
+///     char* str = qk_param_to_string(a);
+///     printf(str);
```
It would be nice to include freeing both the param and, more importantly, freeing the string here via `qk_str_free` to teach users they how to free strings obtained from Qiskit's C API

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+    let str = match expr {
+        Param::ParameterExpression(expr) => expr.to_string(),
+        Param::Float(f) => f.to_string(),
+        Param::Obj(o) => o.to_string(),
```
This case does is invalid, it would require the Python GIL, right?
```suggestion
        Param::Obj(o) => panic!("Param::Obj is not supported in the C API"),
```

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+///
+/// # Safety
+///
+/// lhs and rhs should be valid pointer to parameter
```
Also `out` needs to be a valid pointer -- and all pointers must be non-null too

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+}
+
+/// @ingroup QkParam
+/// calculate exponent of a Param.
```
This sounds a bit unclear to me, how about
```suggestion
/// Apply the exponential function to a ``QkParam``.
```

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+}
+
+/// @ingroup QkParam
+/// calculate absolute of a Param.
```
```suggestion
/// Calculate the absolute value of a ``QkParam``.
```

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+            *out = Param::ParameterExpression(Arc::new(expr.sign()));
+            ExitCode::Success
+        }
+        Param::Float(f) => {
```
You could use `f.signum()` here I think

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+    }
+}
+
+/// @ingroup QkParam
```
I think we should just remove all complex features for parameters until we need them here, otherwise we have a feature in the API that cannot be used but we have to maintain 🤔 or is there a use-case where we would need complex parameters in the C API?

doichanj replied:
I think if we can have complex number, we can get unitary matrix in unitary gate from C-API (currently we can not get unitary matrix by `qk_circuit_get_instruction`)

Cryoris replied:
That would require updating `QkInstruction.params` to `QkParam*`, which is currently `double*` -- but this update likely makes sense anyways since we'll also want to query parameter expressions. Ok then let's keep the complex handling 👍🏻

Cryoris commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1120 @@
+            *out = *f;
+            true
+        }
+        // PyObject should be supported in C-API ?
```
no 🙂

Cryoris reviewed (commented):
I haven't looked at bind/subs in detail yet, but the rest looks pretty good to me. The docstrings need some more care to ensure they contain complete sentences, that the C API objects are named correctly (and annotated with quotes), and to show fully working examples.

I was wondering whether we should introduce a `qk_param_zero()` function for convenience that returns a `Param::Float(0.)`, which we encourage users to use as the `out` variable in the functions?

doichanj committed:
Update crates/cext/src/param.rs

Co-authored-by: Julien Gacon <gaconju@gmail.com>

doichanj committed:
Update crates/cext/src/param.rs

Co-authored-by: Julien Gacon <gaconju@gmail.com>

doichanj committed:
Update crates/cext/src/param.rs

Co-authored-by: Julien Gacon <gaconju@gmail.com>

doichanj committed:
Update crates/cext/src/param.rs

Co-authored-by: Julien Gacon <gaconju@gmail.com>

doichanj committed:
Update crates/cext/src/param.rs

Co-authored-by: Julien Gacon <gaconju@gmail.com>

doichanj committed:
Update crates/cext/src/param.rs

Co-authored-by: Julien Gacon <gaconju@gmail.com>

doichanj committed:
Update crates/cext/src/param.rs

Co-authored-by: Julien Gacon <gaconju@gmail.com>

doichanj committed:
Merge branch 'main' into parameter-c-api

doichanj committed:
add qk_param_zero(), rename qk_param_new to qk_param_new_symbol

doichanj committed:
remove unsafe from qk_param_zero

doichanj committed:
Merge remote-tracking branch 'upstream/main' into parameter-c-api

doichanj committed:
Merge remote-tracking branch 'upstream/main' into parameter-c-api

Cryoris committed:
rename ``qk_param_str`` for consistency

compare to ``qk_obs_str``

Cryoris committed:
Cleanup tests

- use goto for shorter code
- handle QkExitCode from arithmetic functions
- add docstrings

Cryoris committed:
Doc chores

- proper code highlighting
- fix C code sections
- fix safety sections
- update exit code docs

Cryoris committed:
Docs & small improvements for bind/subs

Cryoris committed:
More tests

Cryoris reviewed (commented):
This LGTM as is, though I think it's adding some functionality we don't quite need yet -- which makes it more difficult to get the interface right, as we don't have many practical examples. But given that it's working now and the interface seems good (and felt sensible when writing the tests) I'm happy to merge as is. 

I did add a few commits so @doichanj please double check that you agree before we merge 🙂

doichanj committed:
Merge branch 'main' into parameter-c-api

doichanj commented:
LGTM, thank you for improving docs and tests

Cryoris commented on releasenotes/notes/c-api_param-4f6cd2245a952f66.yaml:
```diff
@@ -0,0 +1,7 @@
+    Introducing Param C-API to support parameterized circuits on C-API.
+    `QkParam` represents symbolic equations through `qk_param_*` functions.
+    Param C-API supports, add/sub/mul/div/pow binary operations and
+    sin/cos/tan/asin/acos/atan/log/exp/abs/sign/neg/conjugate unary operations.
```
```suggestion
    The C API now exposes ``QkParam`` to represent gate parameters, including
    symbols and parameterized expressions. The ``QkParam`` interface itself 
    supports symbolic calculations including standard arithmetic operations,
    trigonometric functions, and binding/substituting values. For example:
    
    .. code-block:: c
    
        QkParam *x = qk_symbol_new("x"); // a symbol called "x"
        QkParam *y = qk_symbol_new("y"); // a symbol called "y"
        QkParam *val = qk_symbol_from_double(2.0); // the value 2.0
        QkParam *out = qk_symbol_zero(); // to store the final expression
        
        // build 2 sin(x + y)
        qk_param_add(out, x, y); // out = x + y
        qk_param_sin(out, out); // out = sin(out)
        qk_param_mul(out, out, val); // out = out * 2
        
        qk_param_free(x);
        qk_param_free(y);
        qk_param_free(val);
        qk_param_free(out);
 
```

Cryoris committed:
Update reno

Cryoris commented:
This LGTM, let's merge to unblock the further workstream on parameters in the C API and the C circuit library. In case we need to adjust the interface there's still time for v2.3 but since there wasn't any further feedback I think we're good 👍🏻

jakelishman commented on crates/cext/src/param.rs:
````diff
@@ -0,0 +1,1293 @@
+/// QkParam *a = qk_param_new_symbol("a");
+/// ```
+///
+/// # Safety
+///
+/// The `name` parameter must be a pointer to memory that contains a valid
+/// nul terminator at the end of the string. It also must be valid for reads of
+/// bytes up to and including the nul terminator.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_param_new_symbol(name: *const c_char) -> *mut Param {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let name = unsafe { CStr::from_ptr(name).to_str().unwrap() };
````
Either the "safety" or the general docs should comment that the string must be valid UTF-8 (plus a nul terminator) as well.

Cryoris replied:
This is a copy of the `char*` docs from the circuit -- I'll add a follow up PR to change this consistently, such that we don't start touching the `circuit.rs` file here 👍🏻

jakelishman commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1293 @@
+///
+/// The behavior is undefined if ``param`` is not a valid pointer to a non-null ``QkParam``.
+///
+/// The string must not be freed with the normal C free, you must use ``qk_str_free`` to
+/// free the memory consumed by the String. Not calling ``qk_str_free`` will lead to a
+/// memory leak.
+///
+/// Do not change the length of the string after it's returned (by writing a nul byte somewhere
+/// inside the string or removing the final one), although values can be mutated.
```
This comment shouldn't be inside the "Safety" section - it should probably be higher up.

Cryoris replied:
Same here -- I'll include this in the PR to ensure it's consistent throughout the C API

jakelishman commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1293 @@
+            ExitCode::Success
+        }
+        Err(_) => ExitCode::ArithmeticError,
+    }
```
Seems slightly odd that we allocate a `String` for the error just to throw it away again immediately.  This is probably illustrating an API problem in `ParameterError`, though, not this function.  Including an always-allocated/calculated error message inside an `Error` type precludes code like this from handling the error in a different way - this is way `derive(Error)` _doesn't_ have everything take a `String` automatically, but instead takes instructions on how to produce a `String`.  (whereas `anyhow`, whose errors are more likely expected to be always abortive, has an API that makes it easier to do things with string allocations)

jakelishman commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1293 @@
+/// pointer to a ``QkParam``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_param_add(
+    out: *mut Param,
+    lhs: *const Param,
+    rhs: *const Param,
+) -> ExitCode {
```
This API requires the user to allocate a `Param` for the `out` slot, which is _always_ overwritten and immediately freed.

An alternative would be to take `out: *mut *mut Param`, so that the _outer_ pointer needs to be aligned/non-null, but it just needs to point to a space to store the new pointer.  You'd call it from C like
```c
// no need to allocate anything into this - it'll never be turned into a
// reference or have its value read
QkParam *out;

QkParam *lhs = qk_param_new_symbol("a");
QkParam *rhs = qk_param_new_symbol("b");

// note the extra `&` on `out`.
if (qk_param_add(&out, lhs, rhs)) {
  // ... handle error ...
}
```

I haven't thought that through 100% of the way - I'm not _completely_ certain that it's more efficient this way round - did you consider this version?

Cryoris replied:
We didn't! What I was going for with the current interface was the analog to something like
```c
Obj *obj = malloc(sizeof(Obj));
do_something_to_obj(obj);
```
such that the user explicitly has to deal with the allocation. I'm not sure which is the more idiomatic version in C...

jakelishman replied:
The reason I queried it is that I'm not certain if you're deriving any benefit from the original allocation; it's always dropped without being read.  That said, as long as the original allocation is just a `Param::Float(0.0)` value, then I'm not sure you're actually losing anything, since in that case you _are_ able to re-use the `Box` space without loss.

This is mostly what I was questioning - I hadn't thought through all the possible uses, and which ones correspond to more/less allocations in the end.  It may well be that your existing way does that, but I'm not 100% sure.

jakelishman commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1293 @@
+        Ok(ret) => {
+            *out = Param::ParameterExpression(Arc::new(ret));
+            ExitCode::Success
+        }
```
This seems to lift `f64 + f64` into the `ParameterExpression` variant.  Is there a reason for that, instead of leaving it as `f64`?

Cryoris replied:
Good point, this should remain `f64` if possible

jakelishman commented on crates/cext/src/param.rs:
Given this is intended for low-level access, should we consider exposing the internal UUID fields of symbols, and/or provide a way to construct a new vresion of a parameter with a particular UUID?

Cryoris replied:
yeah that's a good point -- I also wanted to expose the index in a follow up

jakelishman commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1293 @@
+        // If the input is not parameterized, return a copy.
+        *out = src.clone();
+        ExitCode::Success
+    }
```
1. this might be easier to read if it's turned into a `let Param::ParameterExpression(expr) = src else { *out = src.clone(); return ...; };` rather than a giant `if let`
2. I don't have any concrete suggestions right now, but the interface feels a bit clunky, and certainly feels very expensive if you want to bind parameters in a loop, especially if you want to bind lots of expressions with a lookup table, of which each expression only involves a couple of parameters (since we allocate and traverse a `HashMap` on every call).  Some of that trouble comes from the existing Rust-space `ParameterExpression`, but this C interface is compounding the problem.

Cryoris replied:
1. nice 👍🏻 
2. I agree -- this is part of what I meant with getting the interfaces right without explicit use cases. At the same time I don't think we need to be too worried about this interface here and focus mainly on the circuit parameter assignment. I've also contemplated whether we just want to remove bind/assign for now until we need them and know how we want to use them..

jakelishman commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1293 @@
+    keys: *const *const Param,
+    values: *const f64,
+    num: usize,
+) -> ExitCode {
```
Doesn't need to be this PR, but I suspect we'll want something like `bind_all` in the future (returning an `f64` directly) - the same motivation in Python space will apply here.

jakelishman commented on crates/cext/src/param.rs:
```diff
@@ -0,0 +1,1293 @@
+/// if ``param`` is not a valid, non-null pointer to a ``QkParam``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_param_as_real(out: *mut f64, param: *const Param) -> bool {
```
Did you consider returning a sentinel `f64` in the case of failure, such as a `NaN`, rather than a `bool`?

Cryoris replied:
I don't think so, but that surely seems more consistent with the existing interface, I'll update that

jakelishman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -58,6 +58,8 @@ pub enum ParameterError {
     #[error("Derivative not supported on expression: {0}")]
     DerivativeNotSupported(String),
+    #[error("Operator {0} is not supported.")]
+    OperatorNotSupported(String),
```
Oh, I commented above about this variant being strange by requiring an always-allocated `String`, but I assumed it was pre-existing.  I'd suggest this message is reworked to not required a heap allocation simply to create the error, _especially_ because the only place it's used doesn't even look at the allocation.

Cryoris replied:
I rewrote the logic and this variant is no longer needed 🙂

jakelishman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -189,6 +191,17 @@ impl ParameterExpression {
         }
     }
 
+    /// Try casting to a [Symbol], returning a reference.
+    ///
+    /// This only succeeds if the underlying expression is, in fact, only a symbol.
+    pub fn try_to_symbol_ref(&self) -> Result<&Symbol, ParameterError> {
+        if let SymbolExpr::Symbol(symbol) = &self.expr {
+            Ok(symbol.as_ref())
+        } else {
+            Err(ParameterError::NotASymbol)
+        }
+    }
```
Typically in the Rust stdlib, you wouldn't return a `Result` here, you'd return an `Option<&Symbol>`; there's no information actually carried by the `Err` variant, but this API actually makes it seem like there _is_.

Cryoris replied:
This is mirroring the existing `try_to_symbol -> Result<Symbol, ParameterError>` -- I'll open a follow up PR to address this consistently in both methods

jakelishman commented on releasenotes/notes/c-api_param-4f6cd2245a952f66.yaml:
```diff
@@ -0,0 +1,25 @@
+    The C API now exposes ``QkParam`` to represent gate parameters, including
+    symbols and parameterized expressions. The ``QkParam`` interface itself 
+    supports symbolic calculations including standard arithmetic operations,
+    trigonometric functions, and binding/substituting values. For example:
```
This shouldn't be an upgrade note, it should be a new-feature note.

Cryoris replied:
whoops I missed that -- thanks!

jakelishman commented on test/c/test_param.c:
```diff
@@ -0,0 +1,617 @@
+/**
+ * Test creating a new free symbol and check the name.
+ */
+int test_param_new(void) {
```
Can we make all the `test_*` functions (except the main driver) `static int` instead for the reasons given in #15050?

jakelishman commented on test/c/test_param.c:
```diff
@@ -0,0 +1,617 @@
+    if (real == NULL) {
+        return EqualityError;
+    }
+
+    QkComplex64 c = {0.4, 1.4};
+    QkParam *cmplx = qk_param_from_complex(c);
+    if (cmplx == NULL) {
+        return EqualityError;
+    }
```
This leaks `real`.

jakelishman commented on test/c/test_param.c:
```diff
@@ -0,0 +1,617 @@
+    qk_param_free(p);
+
+    if (strcmp(str, "a") != 0) {
+        printf("The parameter %s is not a\n", str);
```
I can't remember what we do in other tests, but I might expect this to `fprintf(stderr, ...)`, not print to `stdout`.

Cryoris replied:
We always use `printf` actually, but `fprintf(stderr,..)` would actually make more sense since that's also the stream that the success/failure messages are printed to! Another follow up 😄

jakelishman commented on test/c/test_param.c:
```diff
@@ -0,0 +1,617 @@
+    if (fabs(cmplx_out - 1.0) > 1e-10) {
+        printf("Unexpected extracted value in qk_param_as_real.");
+        return EqualityError;
+    }
```
All these guys are missing terminal newlines.

jakelishman commented on test/c/test_param.c:
```diff
@@ -0,0 +1,617 @@
+    }
+
+    str = qk_param_str(ret);
+    if (strcmp(str, "a + b") != 0) {
+        printf("qk_param_add %s is not a + b\n", str);
+        result = EqualityError;
+        goto cleanup_str;
+    }
```
If this test were to fail, it may well be because `str` is corrupted, so trying to print it out without any sort of limitation on the read from ti might not be the best error message.

Cryoris replied:
I updated this to truncate after 10 characters and print `something [...]` like that if it exceeds 10 chars.

jakelishman reviewed (commented):
This wasn't the most in-depth review - I just scanned over it at Julien's request and left a few comments here or there.  I'll generally defer to Julien on what you want to do about the API / if you want to merge this sooner and follow up, or what.

A note that I put a couple of times, but bears another mention: `printf` doesn't include a newline - we use it a lot in the tests, and only occasionally remember to put a trailing newline.

Cryoris committed:
Jake's review comments - light edition

the others are for follow ups :)

Cryoris committed:
Merge branch 'parameter-c-api' of github.com:doichanj/qiskit into parameter-c-api

Cryoris committed:
drop bind/subs for now

there are open discussions on what the right interface to reduce the allocations in practice. Let's unblock parameters while we have this discussion

Cryoris reviewed (dismissed):
LGTM, thanks for the effort @doichanj! After some discussions with Jake, we decided to postpone assign/bind to get the interface right, but still unblock parameters for the C API. I'll also wait for Jake's approval for the changes he requested, since I now also touched this PR a bit.

jakelishman commented on crates/cext/src/exit_codes.rs:
```diff
@@ -96,3 +97,9 @@ impl From<TargetError> for ExitCode {
     }
 }
+
+impl From<ParameterError> for ExitCode {
+    fn from(_value: ParameterError) -> Self {
+        ExitCode::ArithmeticError
+    }
+}
```
This feels like it's making a few too many assumptions now without a `match` statement, but I'm fine if you want to take a note to change the situation later.

jakelishman commented on crates/cext/src/param.rs:
```diff
@@ -478,23 +478,23 @@ pub unsafe extern "C" fn qk_param_pow(
-        (Param::ParameterExpression(base), Param::Float(pow)) => {
-            base.pow(&ParameterExpression::from_f64(*pow))
-        }
+        (Param::ParameterExpression(base), Param::ParameterExpression(pow)) => base
+            .pow(pow)
+            .map(|expr| Param::ParameterExpression(expr.into())),
+        (Param::ParameterExpression(base), Param::Float(pow)) => base
+            .pow(&ParameterExpression::from_f64(*pow))
+            .map(|expr| Param::ParameterExpression(expr.into())),
```
All this stuff _might_ be better on `Param` directly, like `add_numeric` (or whatever we ended up calling the global-phase stuff) - this feels like a lot of code to live in the C API crate for something that's not at all specific to the C API.

Happy to follow up on that, though.

Cryoris replied:
I'd also prefer it directly on `Param` I think, though I didn't want to touch this PR too much -- I'll add it to the follow up list 😄

jakelishman commented on releasenotes/notes/c-api_param-4f6cd2245a952f66.yaml:
```diff
@@ -22,4 +22,5 @@ upgrade_c:
- 
\ No newline at end of file
+
+    Note that binding/substituting parameters is not yet supported.
```
I don't think we necessarily need to say this, but I don't feel strongly.

jakelishman commented on test/c/test_param.c:
```diff
@@ -19,16 +19,30 @@
+static int MAX_STR_LEN = 10;
+
+void printf_limited(char *str, bool newline) {
+    if (strlen(str) > (size_t)MAX_STR_LEN) {
```
If `str` isn't trusted or is expected that it might be corrupt, then `strlen(str)` is already invalid because it has to iterate over the entire string so may read from unmapped memory.

Cryoris replied:
Fair point -- gave this another shot at db309d8, taking care of possible invalid characters and fixing the maximum length manually. But we can also just adapt the message to say `"It didn't print <expected thing>"` instead 🙂

jakelishman reviewed (commented):
Mostly fine (following on from my previous comments only), my only real concern is that `printf_limited` is just as invalid as the previous thing.

Cryoris committed:
one more go at safer print

Cryoris committed:
update reno

Cryoris committed:
fix test artifact

Cryoris committed:
Merge branch 'main' into parameter-c-api

Cryoris committed:
fix rust edition 2024

Cryoris committed:
don't use VLA (thanks MSVC)

Cryoris committed:
rm string prints

jakelishman commented on test/c/test_param.c:
```diff
@@ -11,6 +11,7 @@
 // that they have been altered from the originals.
 
 #include "common.h"
+#include <ctype.h>
```
Do we need anything out of `ctype`?

Cryoris replied:
Ah no this is an artifact from the `print_limited` which was using `isprint` from `ctype.h`!

jakelishman reviewed (dismissed):
With the same caveats / follow-ups as before, the bits I reviewed seem ok to move ahead with.

Cryoris committed:
rm ctype
